bigquery/storage_iterator.go

```diff
@@ -0,0 +1,294 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bigquery
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/bigquery/storage/apiv1/storagepb"
+	"github.com/googleapis/gax-go/v2"
+	"golang.org/x/sync/semaphore"
+	"google.golang.org/api/iterator"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// arrowIterator is a raw interface for getting data from Storage Read API
+type arrowIterator struct {
+	done bool
+	errs chan error
+	ctx  context.Context
+
+	schema  Schema
+	decoder *arrowDecoder
+	records chan arrowRecordBatch
+
+	session *readSession
+}
+
+type arrowRecordBatch []byte
+
+func newStorageRowIteratorFromTable(ctx context.Context, table *Table) (*RowIterator, error) {
+	md, err := table.Metadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+	rs, err := table.c.rc.sessionForTable(ctx, table)
+	if err != nil {
+		return nil, err
+	}
+	it, err := newStorageRowIterator(rs, md.NumRows)
+	if err != nil {
+		return nil, err
+	}
+	it.arrowIterator.schema = md.Schema
+	return it, nil
+}
+
+func newStorageRowIteratorFromJob(ctx context.Context, job *Job, totalRows uint64) (*RowIterator, error) {
+	cfg, err := job.Config()
+	if err != nil {
+		return nil, err
+	}
+	qcfg := cfg.(*QueryConfig)
+	if qcfg.Dst == nil {
+		// TODO: script job ?
```

Comment:
This still needs addressed.  Easy to exercise with a test case using multiple statement execution e.g `SELECT 1 as foo; SELECT foo, bar, baz from sometable;`